### PR TITLE
Allow for platform Thread implementation override

### PR DESCRIPTION
**Pull request by lucypero**
_EDIT:_ Implements https://github.com/godotengine/godot-proposals/issues/3277.

### CODE REVIEW

**Review by RandomShaper — APPROVED** [review]
Looks good to me, besides the formatting issues.

File: core/os/thread.h [hunk 1]
```diff
@@ -28,6 +28,9 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+#ifdef PLATFORM_THREAD_OVERRIDE
```
Comment (akien-mga) [inline core/os/thread.h:31]: There should be a comment to explain what this is for.

File: core/os/thread.h [hunk 2]
```diff
@@ -116,3 +119,4 @@ class Thread {
 };
 
 #endif // THREAD_H
+#endif // PLATFORM_THREAD_OVERRIDE
```
Comment (akien-mga) [inline core/os/thread.h:122]: Missing newline at EOF.

**Comment by lucypero** [comment]
Fixed formatting and added comment.

File: core/os/thread.cpp [hunk 3]
```diff
@@ -28,6 +28,10 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+// If PLATFORM_THREAD_OVERRIDE is defined, it will allow the platform to
+//    define its own thread implementation
```
Comment (akien-mga) [inline core/os/thread.cpp:32]:
```suggestion
// If PLATFORM_THREAD_OVERRIDE is defined, it allows a given platform port
// to define its own thread implementation.
```

File: core/os/thread.h [hunk 4]
```diff
@@ -28,6 +28,11 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 
+// If PLATFORM_THREAD_OVERRIDE is defined, it will allow the platform to
+//    define its own thread implementation
```
Comment (akien-mga) [inline core/os/thread.h:32]:
```suggestion
// If PLATFORM_THREAD_OVERRIDE is defined, it allows a given platform port
// to define its own thread implementation.
```

File: core/os/thread.cpp (hunk 3 above)
Comment (akien-mga) [reply core/os/thread.cpp:32]: Actually I guess you can remove this comment, having it once in the header is enough.

**Comment by Listwon** [comment]
@akien-mga isn't this PR against the contribution best practices? Reduz mentioned them recently on Twitter https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html

These changes are useful only in very special cases like console porting. Dangling `#include "platform_thread.h"` without any stub is a bit misleading. If you are porting engine to special platform, you can handle these tiny additions on your private repository.

**Comment by akien-mga** [comment]
Seems good to me, please squash the commits into one after final edits (see [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) for instructions).

**Comment by akien-mga** [comment]
> These changes are useful only in very special cases like console porting. Dangling `#include "platform_thread.h"` without any stub is a bit misleading. If you are porting engine to special platform, you can handle these tiny additions on your private repository.

I don't think console porting is a "very special case". It's something a significant number of users need to do or get access to to publish their games on consoles. Adding a documented opt-in preprocessor define for that doesn't strike me as particularly problematic.

For console porters, or people porting to some niche platforms, that means that they don't need to do a hard fork of the engine to add a platform port (instead, they can just drop their platform code in the `platforms` folder and compile). Whether that's possible in practice, I don't know, but that's how at least some porters used to work significantly in the past.

I don't think this PR goes against the contribution best practices. It doesn't modify everyone's Thread implementation for a niche use case - it adds an entry point to allow defining a custom one, which is not even compiled if not defined.

**Comment by Listwon** [comment]
In that case "platform_thread.h" doesn't make sense anyway, should be stubbed somehow with some informative comments, thread class stubs, asserts, defines. Previously it was virtual class implemented per platform and now it's `#include` of nonexistent header in root path.

**Comment by reduz** [comment]
I would still add a comment in that line of code added like "Overriding the platform implementation is required in some proprietary platforms".

**Comment by akien-mga** [comment]
> In that case "platform_thread.h" doesn't make sense anyway, should be stubbed somehow with some informative comments, thread class stubs, asserts, defines. Previously it was virtual class implemented per platform and now it's `#include` of nonexistent header in root path.

I don't think it's necessary to make a stub for this. Custom implementation just need to match the interface of `core/os/thread.h` which they replace.

However, it's true that enforcing a `"platform_thread.h"` header which may or may not be in the include path is not the most convenient.
It could be `#include PLATFORM_CUSTOM_THREAD_H` which can be `#define`'d to the platform's relevant file path in `platform_config.h`.

**Comment by reduz** [comment]
Agreed, just a single PLATFORM_CUSTOM_THREAD_H define that points to the include is probably enough.

**Comment by lucypero** [comment]
Replaced `PLATFORM_THREAD_OVERRIDE` with `PLATFORM_CUSTOM_THREAD_H`

**Comment by akien-mga** [comment]
Looks great. Just remember to update the `#endif // PLATFORM_THREAD_OVERRIDE` comments in both files, and squash the commits.

**Comment by lucypero** [comment]
> Looks great. Just remember to update the `#endif // PLATFORM_THREAD_OVERRIDE` comments in both files, and squash the commits.

Done.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Cherry-picked for 3.4.

**Pull request merged**